### PR TITLE
clusterversion: introduce 22.1 development versions

### PR DESCRIPTION
**Pull request by celiala**
Shortly after [Creating a release branch](https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/187859111/Creating+a+release+branch) for release-21.2, we'll want to merge PRs as instructed by the [New major version checklist](https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/1522270228/New+major+version+checklist).

This PR is for Step 1b of the [New major version checklist](https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/1522270228/New+major+version+checklist), where we add a cluster version `VersionStart${vNEXT}`.

This is part of #70751, which tracks all the steps relevant to creating a release branch for `${vBRANCH_CUT}` and preparing master for the `${vNEXT}` major version.

Release justification: Non-production code change.
Release note: None

### CODE REVIEW

**Comment by celiala** [comment]
This should not be merged until after Tuesday's branch cut.

File: pkg/clusterversion/key_string.go [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // Code generated by "stringer -type=Key"; DO NOT EDIT.
+// TODO(celia) -- how do I re-autogenerate this file?
```
Comment (celiala) [inline pkg/clusterversion/key_string.go:2]: TODO(celia) - who might know the answer to this?

File: pkg/clusterversion/cockroach_versions.go [hunk 2]
```diff
@@ -507,6 +514,27 @@ var versionsSingleton = keyedVersions{
...
+		// V21_2 is CockroachDB v21.2. It's used for all v21.2.x patch releases.
+		Key:     V21_2,
+		Version: roachpb.Version{Major: 21, Minor: 2},
+	},
+
+	// v21.2 versions.
+	{
+		Key:     Start21_2,
+
+		// TODO(celia) - what should `Internal` be?
```
Comment (celiala) [inline pkg/clusterversion/cockroach_versions.go:527]: Maybe @RaduBerinde ?

File: pkg/clusterversion/cockroach_versions.go [hunk 3]
```diff
@@ -507,6 +514,27 @@ var versionsSingleton = keyedVersions{
...
+		// - 162 <== {160, the last_value} + 2?
+		// - or should I jump up a few numbers (like 200?), to leave space for 21.1.x patch key/values?
+		Version: roachpb.Version{Major: 21, Minor: 1, Internal: 200},
+	},
+
+	// TODO(celia) - should I be making a Start21_2PLUS, similar to
```
Comment (celiala) [inline pkg/clusterversion/cockroach_versions.go:533]: Maybe @RaduBerinde here as well?

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69828)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @celiala, and @irfansharif)

___
*[pkg/clusterversion/cockroach_versions.go, line 477 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69828#-Mj6SkdN8Bvf-JNDPEM2:-Mj6SkdO867RgsgezH_2:b-glitdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/63b7dd4243f01f0079ed091596a70602474eaf79/pkg/clusterversion/cockroach_versions.go#L477)):*
> ```Go
> 	// TODO(celia) - should `Internal` be 1160+2? Or should a use something bigger like 2000?
> 	{
> 		Key:     Start21_2,
> ```

Start22_1
___
*[pkg/clusterversion/cockroach_versions.go, line 478 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69828#-Mj6SlffAuS5Nh-cqGEG:-Mj6SlffAuS5Nh-cqGEH:b59gr9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/63b7dd4243f01f0079ed091596a70602474eaf79/pkg/clusterversion/cockroach_versions.go#L478)):*
> ```Go
> 	{
> 		Key:     Start21_2,
> 		Version: roachpb.Version{Major: 21, Minor: 1, Internal: 2000},
> ```

Major 22 Minor 1. Internal can be (say) 100, we just want to leave a gap in case we need it later. These values are never compared against internal values from a different Major/Minor.
___
*[pkg/clusterversion/cockroach_versions.go, line 481 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69828#-Mj6Stdd1ADZxM3w16Rd:-Mj6Stdd1ADZxM3w16Re:bcsjr3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/63b7dd4243f01f0079ed091596a70602474eaf79/pkg/clusterversion/cockroach_versions.go#L481)):*
> ```Go
> 	},
> 
> 	// TODO(celia) - should I be making a Start21_2PLUS, similar to Start21_1PLUS for serverless?
> ```

No PLUS, ignore that. We didn't end up doing that.
___
*[pkg/clusterversion/key_string.go, line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69828#-MihiIHdcncvOk8EllYe-r2-2:-Mj6TsAqDxLhybEbrTSu:bu1cbkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/742c1c15aded1b721855485e0994f31dad419d36/pkg/clusterversion/key_string.go#L2)):*
<details><summary><i>Previously, celiala wrote…</i></summary><blockquote>

TODO(celia) - who might know the answer to this?
</blockquote></details>

`go generate` in that dir should do it I think. Or `make generate`

File: pkg/clusterversion/cockroach_versions.go [hunk 4]
```diff
@@ -458,6 +465,21 @@ var versionsSingleton = keyedVersions{
...
+	{
+		// V21_2 is CockroachDB v21.2. It's used for all v21.2.x patch releases.
+		Key:     V21_2,
+		Version: roachpb.Version{Major: 21, Minor: 2},
+	},
+
+	// v21.2 versions.
+	// TODO(celia) - should `Internal` be 1160+2? Or should a use something bigger like 2000?
+	{
+		Key:     Start21_2,
```
Comment (irfansharif) [inline pkg/clusterversion/cockroach_versions.go:477]: I think we want to use `Start22_1` here;  Start21_2  is already defined above. Also, the version should be `roachpb.Version{Major: 21, Minor: 2, Internal: 2}`. The +14, +100, and +1000 were due to historical artifacts. Pretty much any number >= 2 would work here, but I think we should start at 2.

File: pkg/clusterversion/cockroach_versions.go [hunk 5]
```diff
@@ -458,6 +465,21 @@ var versionsSingleton = keyedVersions{
...
+		Version: roachpb.Version{Major: 21, Minor: 1, Internal: 2000},
+	},
+
+	// TODO(celia) - should I be making a Start21_2PLUS, similar to Start21_1PLUS for serverless?
```
Comment (irfansharif) [inline pkg/clusterversion/cockroach_versions.go:481]: Don't think so.

**Review by irfansharif — APPROVED** [review]

**Comment by irfansharif** [comment]
(The LGTM was for code changes -- not sure when this PR is supposed to get merged. Looking at the discussion over at https://github.com/cockroachdb/cockroach/pull/69826, doesn't seem like it would be now)

**Comment by celiala** [comment]
TFTR!

> (The LGTM was for code changes -- not sure when this PR is supposed to get merged. Looking at the discussion over at #69826, doesn't seem like it would be now)

Yes, I'll hold off merging this until we've released a beta. I updated the PR description to clarify when I intend to merge - thanks!

**Comment by irfansharif** [comment]
When we merge this, we also want to bump the binaryMinSupportedVersion to point to V21_2 instead.

**Comment by irfansharif** [comment]
https://github.com/cockroachdb/cockroach/blob/6464de2569efeee6c9d0bbbe7bb5be476da0cf03/pkg/clusterversion/cockroach_versions.go#L86-L93 it's kind of tucked away here, is there a TODO list you're maintaining elsewhere Celia?

**Review by celiala — COMMENTED** [review]
> we also want to bump the binaryMinSupportedVersion to point to V21_2 instead.

Ack. Done.


> [the steps in the comments are] kind of tucked away here, is there a TODO list you're maintaining elsewhere Celia?

Ack, good callout!

Yes. We currently have these wiki pages that I've been running through:

- https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/187859111/Creating+a+release+branch
- https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/1522270228/New+major+version+checklist

But with "multiple sources of truth" there's a chance that variations will happen. I have a running list of things to improve[]1 for the next branch cut -- I'll add this "multiple sources of truth" as something to consolidate. Thanks!

[1] https://cockroachlabs.atlassian.net/wiki/spaces/devinf/pages/2218197098/21.2+release+process+notes+a+running+list+retro

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69828#-:-MkOL8Hk2ReiKHRIlnvz:bltwerx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @irfansharif)

___
*[pkg/clusterversion/cockroach_versions.go, line 477 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69828#-MihiIHdcncvOk8EllYd-r4-477:-MkOKD6R1BqjvIoFWbdd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/63b7dd4243f01f0079ed091596a70602474eaf79/pkg/clusterversion/cockroach_versions.go#L477)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

I think we want to use `Start22_1` here;  Start21_2  is already defined above. Also, the version should be `roachpb.Version{Major: 21, Minor: 2, Internal: 2}`. The +14, +100, and +1000 were due to historical artifacts. Pretty much any number >= 2 would work here, but I think we should start at 2.
</blockquote></details>

Done.
___
*[pkg/clusterversion/cockroach_versions.go, line 481 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/69828#-MihiIHdcncvOk8EllYd-r4-481:-MkOKHZ2DLV-zlj4GRka:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/63b7dd4243f01f0079ed091596a70602474eaf79/pkg/clusterversion/cockroach_versions.go#L481)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Don't think so.
</blockquote></details>

Done.

**Comment by celiala** [comment]
🤔 I'm seeing this pattern of error:

```
E211001 16:44:27.695542 51 server/init.go:326  [n?] 18  bootstrap: store &{engine1 false <no-attributes>=<in-mem> 0xc0004a5260}, last used with cockroach version v21.1-1132, is too old for running version v21.1-1132 (which requires data from v21.2 or later)
    testcluster.go:307: unable to bootstrap due to internal error
```

On a handful of tests:
- pkg/ccl/migrationccl/migrationsccl/
	- TestRecordsBasedRegistryMigration
- pkg/migration/migrations/
	- TestDeleteDeprecatedNamespaceDescriptorMigration
	- TestDeleteDeprecatedNamespaceDescriptorMigrationOnlyNamespace2
	- TestDeleteDeprecatedNamespaceDescriptorMigrationNoOp
...
 
It sounds like this needs to pull a `v21.2-x` build, but I'm not sure how these tests get setup, or how to update that..

Ideas?

**Comment by irfansharif** [comment]
So these tests in the `migration{,ccl}` packages override the binary version in order to test the execution of a specific migration association with a specific cluster version.

https://github.com/cockroachdb/cockroach/blob/617d300d909fa808c6b528ace5f3a033895d51d1/pkg/migration/migrations/delete_deprecated_namespace_tabledesc_external_test.go#L43

Given that this PR bumps the minimum supported binary version, for these tests we're in the incorrect state where the binary version is actually less than the minimum supported version (#71684 would've made this clearer). One thing we could do is also these tests override the minimum supported version. I've paged most of this out, but I think the way we'd do it by upgrading all tests that directly touch BinaryVersionOverride to instead using a mocked out clusterversion.Handle instead, threading that in appropriately:

https://github.com/cockroachdb/cockroach/blob/c92936ea2d048b12469ff908688cd41bc4939280/pkg/server/testing_knobs.go#L79-L81


Failed attempt: I tried a quick revision adding a `BinaryMinimumVersionOverride`, a sister testing knob to `BinaryVersionOverride`, but it doesn't play super well with the fact that the cluster version setting is [initialized globally](https://github.com/cockroachdb/cockroach/blob/f3c94ef8645abd1cfb6064dee11c3d2afdb0add3/pkg/clusterversion/clusterversion.go#L51-L60)) using the real min-supported version, which prevents tests from easily overriding the server's view of min-supported versions because it's [checked against the cluster setting](https://github.com/cockroachdb/cockroach/blob/904807640f6a28aa82c5e63f0b62de2dbf27b3a0/pkg/clusterversion/setting.go#L79-L91). At that point we might as well plumb in a testing knob for a clusterversion.Handle instead.

---

@ajwerner, is there another workaround we're missing? @celiala, I'd be ok with skipping the bump of the min supported version but adding a TODO+filing an issue to come back around to it after addressing the above. We'd need to get to it before any 22.1 release.

**Pull request merged**